### PR TITLE
`azurerm_storage_account`/`azurerm_storage_account_blob_properties`/`azurerm_storage_account_queue_properties`/`azurerm_storage_account_share_properties`/`azurerm_storage_account_static_website_properties` - split data plane into separate resources

### LINKED ISSUE #2977: Cannot create azurerm_storage_container in azurerm_storage_account that uses network_rules
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version
Terraform v0.11.11
+ provider.azurerm v1.21.0

### Affected Resource(s)

* `azurerm_storage_account`
* `azurerm_storage_container`

### Terraform Configuration Files

```
resource "azurerm_storage_account" "test-storage-acct" {
  name                     = "${var.prefix}storacct"
  resource_group_name      = "${var.resgroup}"
  location                 = "${var.location}"
  account_tier             = "Standard"
  account_replication_type = "LRS"
  network_rules {
    ip_rules                   = ["aaa.bbb.ccc.ddd/ee"]
    virtual_network_subnet_ids = ["${var.subnetid}"]
  }
}
resource "azurerm_storage_container" "provisioning" {
  name                  = "${var.prefix}-provisioning"
  resource_group_name   = "${var.resgroup}"
  storage_account_name  = "${azurerm_storage_account.test-storage-acct.name}"
  container_access_type = "private"
}
```

### Debug Output

* azurerm_storage_container.provisioning: Error creating container "philtesting1-provisioning" in storage account "philtesting1storacct": storage: service returned error: StatusCode=403, ErrorCode=AuthorizationFailure, ErrorMessage=This request is not authorized to perform this operation.
RequestId:a7f9d2e1-701e-00b3-4e74-cf3b34000000
Time:2019-02-28T14:45:53.7885750Z, RequestInitiated=Thu, 28 Feb 2019 14:45:53 GMT, RequestId=a7f9d2e1-701e-00b3-4e74-cf3b34000000, API Version=, QueryParameterName=, QueryParameterValue=

### Expected Behavior

Container can be created in a storage account that uses network rules

### Actual Behavior

After applying a network_rule to a storage account I cannot provision a container into it.  My public IP is included in the address range specified in the network rule.  I can successfully create the container via the Azure portal

### Steps to Reproduce

1. `terraform apply`

### PR DESCRIPTION
**Pull request by WodansSon**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

:bangbang: **IMPORTAINT** :bangbang:: **This PR must be included into the last v3.x provider before the v4.0 release.**

**NEW FEATURES FLAG:**
* `data_plane_access_on_read_enabled`

**NEW RESOURCES:**
* `azurerm_storage_account_blob_properties`
* `azurerm_storage_account_queue_properties`
* `azurerm_storage_account_share_properties`
* `azurerm_storage_account_static_website_properties`

**EXISTING RESOURCES BEHAVIOR UPDATED:**
* `azurerm_storage_account` - deprecated `blob_properties`, `share_properties`, `queue_properties` and `static_website` code blocks for `create` 

## PR Checklist

- [X] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [X] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [X] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [X] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [X] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [X] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [X] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [X] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [X] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

image

* **NOTE:** The `TestAccStorageAccount_disappears` test case fails in `main` as well. 

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

### FEATURES:

* New Resource: `azurerm_storage_account_blob_properties` [GH-00000]
* New Resource: `azurerm_storage_account_queue_properties` [GH-00000]
* New Resource: `azurerm_storage_account_share_properties` [GH-00000]
* New Resource: `azurerm_storage_account_static_website_properties` [GH-00000]

### ENHANCEMENTS:

* provider: support for the feature flag `data_plane_access_on_read_enabled` [GH-00000]

### DEPRECATIONS:
* `azurerm_storage_account` - deprecated `blob_properties`, `share_properties`, `queue_properties` and `static_website` code blocks for `create` [GH-00000]

This is a (please select all that apply):

- [ ] Bug Fix
- [X] New Feature (ie adding a service, resource, or data source)
- [X] Enhancement
- [X] Breaking Change

## Related Issue(s)
Fixes #2977

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/storage/storage_account_resource.go [hunk 1]
```diff
@@ -1261,6 +1105,178 @@ func resourceStorageAccount() *pluginsdk.Resource {
...
+						Optional:     true,
+						ValidateFunc: validation.StringIsNotEmpty,
+					},
+					"index_document": {
+						Type:         pluginsdk.TypeString,
+						Optional:     true,
+						ValidateFunc: validation.StringIsNotEmpty,
+					},
+				},
+			},
```
Comment (jackofallops) [inline internal/services/storage/storage_account_resource.go:1278]: I think you have an erroneous `,` here... 
```suggestion
			}
```

**Review by jackofallops — DISMISSED** [review]
Hi @WodansSon - As discussed offline, there are some changes required for the newly introduced resources, both that they need to use the TypedSDK and the schema needs breaking out of the `properties` block.  Since we've not finalised the agreed approach for the data plane changes, I've only left a partial review for now until we have test cases / repro's that prove the approach, and demonstrate that customers not experiencing the data plane issues are not adversely affected by the changes. 
I'm unfortunately out for a week on PTO now, so @katbyte will follow up with you offline and I'll pick back up on my return.

**Comment by WodansSon** [comment]
I had to do a `git push origin e_storage_account_dataplane --force` to revert the code back to the `v3.x` provider code base, else customer would not be able to build a private and test the fix locally. I mistakenly pulled the `v4.0` changes into this PR branch with commit `872faac`.

**Comment by WodansSon** [comment]
I have removed the `breaking-change` label from this PR as it is not really a breaking change per se. Pre-existing storage accounts will still be able to be managed as they always have. This change is more a change in behavior of the resource for net new storage accounts.

File: internal/services/storage/storage_account_queue_properties_resource.go [hunk 2]
```diff
@@ -0,0 +1,656 @@
...
+									"version": {
+										Type:         pluginsdk.TypeString,
+										Required:     true,
+										ValidateFunc: validation.StringIsNotEmpty,
+									},
+									// TODO 4.0: Remove this property and determine whether to enable based on existence of the out side block.
+									"enabled": {
+										Type:     pluginsdk.TypeBool,
+										Required: true,
+									},
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply internal/services/storage/storage_account_queue_properties_resource.go:81]: The comment is still valid as this PR is for the v3.x provider, when I create the PR for the v4.0 provider I will implement this change.

File: internal/services/storage/storage_account_queue_properties_resource.go [hunk 3]
```diff
@@ -0,0 +1,656 @@
...
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply internal/services/storage/storage_account_queue_properties_resource.go:144]: As above.

File: internal/services/storage/storage_account_queue_properties_resource.go [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/timeouts"
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+	"github.com/tombuildsstuff/giovanni/storage/2023-11-03/queue/queues"
+)
+
+var storageAccountQueuePropertiesResourceName = "azurerm_storage_account_queue_properties"
+
+func resourceStorageAccountQueueProperties() *pluginsdk.Resource {
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply internal/services/storage/storage_account_queue_properties_resource.go:30]: As discussed offline, I am happy to convert this to a typedSDK resource for the v4.0 version of this resource, at this point in time, I don't see the value in doing this at the moment...

File: internal/services/storage/storage_account_queue_properties_resource.go [hunk 5]
```diff
@@ -0,0 +1,656 @@
...
+
+		Schema: map[string]*pluginsdk.Schema{
+			"storage_account_id": {
+				Type:         pluginsdk.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: commonids.ValidateStorageAccountID,
+			},
+
+			"properties": {
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply internal/services/storage/storage_account_queue_properties_resource.go:57]: Again, I don't disagree, I will be flattening this for the v4.0 version of this resource, as for now I feel this is fine...

File: internal/services/storage/storage_account_static_website_properties_resource.go [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+
+		Schema: map[string]*pluginsdk.Schema{
+			"storage_account_id": {
+				Type:         pluginsdk.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: commonids.ValidateStorageAccountID,
+			},
+
+			"properties": {
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply internal/services/storage/storage_account_static_website_properties_resource.go:55]: Again, I don't disagree, I will be flattening this for the v4.0 version of this resource, as for now I feel this is fine...

**Pull request closed without merging**